src/Http/Http.Extensions/src/RequestDelegateFactory.cs

```diff
@@ -516,21 +535,67 @@ private static Expression GetValueFromProperty(Expression sourceExpression, stri
+        }
+
         private static Expression BindParameterFromValue(ParameterInfo parameter, Expression valueExpression, FactoryContext factoryContext)
         {
+            var nullability = nullabilityContext.Create(parameter);
+            var isOptional = parameter.HasDefaultValue || nullability.ReadState == NullabilityState.Nullable;
+
             if (parameter.ParameterType == typeof(string))
             {
-                if (!parameter.HasDefaultValue)
+                factoryContext.UsingTempSourceString = true;
+
+                if (!isOptional)
                 {
-                    return valueExpression;
+                    // The following is produced if the parameter is required:
+                    //
+                    // tempSourceString = httpContext.RouteValue["param1"] ?? httpContext.Query["param1"];
+                    // if (tempSourceString == null)
+                    // {
+                    //      wasTryParseFailure = true;
+                    //      Log.RequiredParameterNotProvided(httpContext, "Int32", "param1");
+                    // }
+                    var checkRequiredStringParameterBlock = Expression.Block(
+                        Expression.Assign(TempSourceStringExpr, valueExpression),
+                        Expression.IfThen(TempSourceStringNullExpr,
+                            Expression.Block(
+                                Expression.Assign(WasTryParseFailureExpr, Expression.Constant(true)),
+                                Expression.Call(LogRequiredParameterNotProvidedMethod, 
+                                    HttpContextExpr, Expression.Constant(parameter.ParameterType.Name), Expression.Constant(parameter.Name))
+                            )
+                        )
+                    );
+
+                    factoryContext.TryParseParams.Add((TempSourceStringExpr, checkRequiredStringParameterBlock));
+                    return Expression.Block(TempSourceStringExpr);
                 }
 
-                factoryContext.UsingTempSourceString = true;
+                // Allow nullable parameters that don't have a default value
+                if (nullability.ReadState == NullabilityState.Nullable && !parameter.HasDefaultValue)
+                {
+                    return Expression.Block(Expression.Assign(TempSourceStringExpr, valueExpression));
+                }
+
+                // The following is produced if the parameter is optional. Note that we convert the
+                // default value to the target ParameterType to address scenarios where the user is
+                // is setting null as the default value in a context where nullability is disabled.
+                //
+                // tempSourceString = httpContext.RouteValue["param1"] ?? httpContext.Query["param1"];
+                // tempSourceString != null ? tempSourceString : Convert("3", Int32)
```

Comment:
I'm not sure this makes sense. `tempSourceString` has to be a `string` right? Do you have a test case this new cast fixes? Maybe if a `string` param had a non-string default value that had an implicit conversion?

Reply:
I'll fix the example in the code comment. This is specifically to convert `null` to the reference type in the scenario where a user does something like: `(string foo = null) => ""`